### PR TITLE
Use SquareSpace LESS compiler instead of less.js

### PR DESCRIPTION
**Pull request by Aeon**
See https://github.com/Squarespace/less-compiler

Benefits:
 1. faster than lessjs
 2. supports squarespace-specific LESS features
   a. SquareSpace "tweaks" concept: http://developers.squarespace.com/style-editor/

   b. escaping variables to be used as selectors (see http://stackoverflow.com/questions/13553669/less-css-escape-interpolated-string-for-multiple-selectors) - see this example from `mobile-nav.less`

      ```less
    (~"@{mobileEl}") { ... }

### CODE REVIEW

**Comment by Aeon** [comment]
So, here's a commit that hooks sqs-less into the build process: 1903a7e.

As for the undefined LESS vars and mixins, yeah, I'm hitting that, but that's going to be an issue no matter what compiler you use, depending on your template.

Here's my preliminary research notes though, maybe we can figure something out:

After poking around the admin UI, I notice that it has template tweak metadata API endpoint like: 

https://[domain].squarespace.com/api/template/GetTemplateTweakSettings

that serves a json file with a whole bunch of useful stuff. From what I can see: 
 - `tweakIdMap` contains all the var references
 - `rawLess` contains all the generated LESS code, including both our template stuff AND squarespace-internal LESS markup with mixins and such goodies
 - `tweakLess` contains only the tweak-specific LESS code
 - `tweakValues` contains all the variable values that user set from admin tool
 - `presets.templateTweaks` contains all the var default values, to be used as fallback I guess
 
Plan of attack:
  - grab latest template vals from site on server start
  - unescape rawless
    - replace \\n with \n
    - replace \" with "
  - dump resulting rawLess to styles/raw.less
  - iterate over tweakIdMap, and generate a LESS variable for each
    - use value from tweakValues if set
    - use value from presets.templateTweaks if not
  - dump tweak vars to tweakValues.less
  - generate style.less that contains:
     - reset.css
     - tweakValues.less
     - rawLess.less
     - everything else in styles subdirectory that user defined in their template, *.less

The main issue I can see is that rawLess key contains both squarespace internal, and template specific LESS code. But maybe with some clever slicing, we can figure out where squarespace internal code starts and ends, and extract just that part, so the remainder is not duplicated when generated from the rest of template LESS files.

File: squarespace-template.js [hunk 1]
```diff
@@ -795,51 +797,34 @@ compilePages = function ( cb ) {
...
-                if ( !exists ) {
-                    read();
+    fs.writeFile( path.join(directories.styles, 'style.less') , styles.join("\n") , function(err) {
+        if(err) {
+            return sqsLogger.log( "error", err);
+        }
 
+        execFile(
+            "/usr/local/bin/sqs-lessc",
+            ['style.less', 'style.css'],
```
Comment (Aeon) [inline squarespace-template.js]: oh, also, this will render to file, so I dunno how well that will play with livereload stuff.

**Comment by Aeon** [comment]
also, this might help with #123

**Comment by kitajchuk** [comment]
The endpoint for `GetTemplateTweakSettings` looks promising. I've been playing around with the lessc compiler as well and it seems promising beyond the system imports. The ones I know are these

- sqs-grid-breaker
- sqs-ui-font
- social-icon-font ( deprecated )

I have a local branch as of right now that I have been playing with this stuff in. Your outlined plan of attack generally covers the gist of what would need to be done.

It's a bit more involved since implementing the new compiler and resolving any other of the Less issues shouldn't require the usage of the node server to change in any way.

I'm going to avoid merging your PR since the full-scale implementation of this is quite large. The middleware would require a small update as well and I still see some unkowns.

The following is an updated scope of this implementation as I see it.

1. Update `node-squarespace-middleware` to consume data from `api/template/GetTemplateTweakSettings` when logging in. This will amend the [getAPIData](https://github.com/NodeSquarespace/node-squarespace-middleware/blob/master/squarespace-middleware.js#L156) method which pre-fetches necessary schemas for the node server to operate effectively.
2. Implement the `lessc` compiler as an executable within the `node-squarespace-server` module at `bin/lessc`. This way the module ships with it and developer usage of the server doesn't change.
3. Utilize `rawLess` from the `GetTemplateTweakSettings` information, parsing out only the Squarespace System portion ( not sure how hard/easy this will be ). Since `rawLess` retains comments still, it could be as easy as having folks use a comment like `/* node-squarespace-server hook */` as the first line of their styles to parse it all out.
4. Parse the remote `site.css` absolute `<link />` out of rendered local template headers so only local styles render locally.
5. Swap the current node Less compiler for the `lessc` compiler using child_process.exec.
6. Use the `GetTemplateTweakSettings` information to create a tweak variables Less stylesheet for use with the `lessc` compiler.
7. Compile the styles in order of `template.conf:stylesheets`, prepending `sqs-system-raw.less` and `sqs-tweak-variables.less`. The `reset.css` file will be accounted for as well if applicable. I'll note here that the system imports mentioned above that get used in template Less files would still not work correctly. The only thing I can think is that as long as you are using them, the `rawLess` would potentially include those files' imported contents - though that doesn't seem likely.

I think this generally makes sense. @Aeon if you have any thoughts/insights on the system imports, please share them here since that seems like the biggest hurdle still in this whole process.

**Comment by Aeon** [comment]
@kitajchuk I haven't really spent much time digging through it, but my impression was that the definitions of `sqs-grid-breaker`, `sqs-ui-font`, etc, were all included in rawLess key. If that's the case, it should be fine to either strip those imports from templates that require them, or try to extract them from rawLess into separate files as needed. Maybe SQS dev support can provide some insight too, I haven't tried filing a ticket with them.

**Comment by Aeon** [comment]
Also, I guess I shouldn't have converted the issue into a pull request, this was really just proof of concept and I wasn't expecting it to be merged anyway. Feel free to close it and open a new issue to track this project. Sorry for the inconvenience.

**Comment by kitajchuk** [comment]
@Aeon - That's what I was hoping for with the rawLess. I think some playing around may uncover the nature there. No worries about the PR, this is a good conversation as its been the biggest issue on my mind about the whole project in a lot of ways.

No worries about the PR. I'm going to open an actual issue and start it with my last notes. This will be a good way to track against all the work :D

Thanks again for your support!

**Pull request closed without merging**